oeps/oep-0050-hooks-extension-framework.rst

```diff
@@ -0,0 +1,235 @@
+   * - Authors
+     - Felipe Montoya <felipe.montoya@edunext.co>
+   * - Arbiter
+     - Nizar Mahmoud <nizar@opencraft.com>
+   * - Status
+     - Draft
+   * - Type
+     - Architecture
+   * - Created
+     - 2021-02-04
+   * - Review Period
+     - < TBD >
+
+Abstract
+========
+
+To sustain the growth of the Open edX ecosystem, the business rules of the
+platform must be open for extension following the open-closed principle. This
+OEP specifies a solution using what is commonly referred to as hooking. That is
+the definition of a stable set of places in the code (triggers) where functions
+defined via plugins can take place.
+
+
+Motivation
+==========
+
+Following the open-closed principle, it would be desirable for the open edX
+project to allow the extension points (great `summary of the current options`_)
+to modify as much of the code paths that define the business rules of an
+open edX instance in a way that is consistent and repeatable.
+
+Since the creation of the plugin framework different initiatives in the
+community have demonstrated success by adopting this pattern. This OEP extends
+the range of possible extensions that the patter can hold sustainably.
+
+Adopters of the Open edX technology as a way of delivering online courses very
+often require that certain changes are made to the codebase by changing the
+edx-platform core. This proposal aims to be a solution to organize the majority
+of the changes that the authors have encountered after years of work extending
+the core in a way that is more compatible with the community at large, it is
+easier to reuse and far easier to migrate to newer versions of the codebase.
+
+This proposal draws inspiration from the cumulative experience of eduNEXT as the
+lead author, other members of the community, and from one of the largest
+extendable platforms for the web today for many years already, WordPress.
+
+.. _summary of the current options: https://github.com/edx/edx-platform/blob/master/docs/guides/extension_points.rst
+
+
+Specification
+=============
+
+The core of this proposed framework is the definition of a stable list of places
+in the life cycle of the learners registration, login, enrollment, course usage,
+and certification. Conversely, the authors life cycle will have a list of places
+during the creation and publication of courses. These places, called triggers
+will be written to the edx-platform code.
+Now, anyone that wants to extend the functionality of the platform at one of
+those places, can define a function, called an action or filter, to be called
+when the trigger is executed. The sum of a trigger and action will be referred
+to as a hook.
+
+Triggers, broadly fit into two categories:
+
+* Triggers that allow actions called at this time to affect the application flow
+  via their results. In this case, the actions called at the time will be called
+  filters.
+* Triggers that will not allow the application flow to be altered by the
+  actions called.
```

Comment:
[puzzled] This definition of `filters` versus `actions` doesn't match what is written in the Wordpress docs.

Filters do not alter the application flow. Instead, they alter the data that is used within the application flow.
Actions halt the application flow, execute the action, and then return back to the normal flow.

What am I missing in my understanding of this proposal?

Reply:
This comment made me clarify a lot of the wording. First I removed "trigger" and "action" from the OEP since it was making it even more complicated. I settled on events and filters (both are kinds of hooks).

I also removed the specific link to the wordpress docs because this proposal is not to copy what Wordpress does, but to be inspired by it.

I clarified the events vs filters part at: https://github.com/edx/open-edx-proposals/pull/184/files#diff-f28146adea40eb552912482a8b24a19107252beba3e25295413bcd5ef8927628R71-R78